File: scala/src/test/scala/BinarySearchSpec.scala
import org.scalatest._

import algorithm._

class BinarySearchSpec extends FlatSpec with Matchers {

  trait Fixture {
    val a = Seq(1, 2, 3, 4)
  }

  "BinarySearch Algorithms" should "lowerBound OK" in new Fixture {
    BinarySearch.lowerBound(a, 3) should be(2)
    BinarySearch.lowerBound(a, 1) should be(0)
    BinarySearch.lowerBound(a, 5) should be(4)
  }

  it should "upperBound OK" in new Fixture {
    BinarySearch.upperBound(a, 3) should be(3)
    BinarySearch.upperBound(a, 4) should be(4)
  }

  it should "binarySearch OK" in new Fixture {
    BinarySearch.binarySearch(a, 3) should be(true)
    BinarySearch.binarySearch(a, 0) should be(false)
    BinarySearch.binarySearch(a, 5) should be(false)
  }

  it should "equalRange OK" in new Fixture {
    BinarySearch.equalRange(a, 3) should be((2, 3))
  }
}

File: typescript/test/binarySearchSpec.js
var should = require('chai').should();
var Algos = require('../dist/binarySearch.js');

var Ary = function() {
  return [1, 2, 3, 4];
};

describe('Binary search algorithms', function(){

  it('lowerBound', function(){
    Algos.lowerBound(Ary(), 3).should.be.equal(2);
    Algos.lowerBound(Ary(), 2.5).should.be.equal(2);
    Algos.lowerBound(Ary(), 1).should.be.equal(0);
    Algos.lowerBound(Ary(), 5).should.be.equal(4);
  });

  it('upperBound', function(){
    Algos.upperBound(Ary(), 3).should.be.equal(3);
    Algos.upperBound(Ary(), 2.5).should.be.equal(2);
    Algos.upperBound(Ary(), 4).should.be.equal(4);
  });

  it('binarySearch', function(){
    Algos.binarySearch(Ary(), 3).should.be.equal(true);
    Algos.binarySearch(Ary(), 0).should.be.equal(false);
    Algos.binarySearch(Ary(), 5).should.be.equal(false);
  });

  it('equalRange', function(){
    Algos.equalRange(Ary(), 3).lowerBound.should.be.equal(2);
    Algos.equalRange(Ary(), 3).upperBound.should.be.equal(3);
    Algos.equalRange(Ary(), 2.5).lowerBound.should.be.equal(2);
    Algos.equalRange(Ary(), 2.5).upperBound.should.be.equal(2);
  });
});
